testcase/cross-region/tso.go

```diff
@@ -0,0 +1,209 @@
+package crossregion
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"sync"
+
+	"github.com/pingcap/log"
+	"go.uber.org/zap"
+
+	util2 "github.com/pingcap/tipocket/testcase/cross-region/pkg/util"
+)
+
+func (c *crossRegionClient) testTSO(ctx context.Context) error {
+	if err := c.requestTSOs(ctx); err != nil {
+		return err
+	}
+	log.Info("start to transfer Leader")
+	if err := c.transferLeader(); err != nil {
+		return err
+	}
+	log.Info("leader transfer committed")
+	if err := c.waitLeaderReady(); err != nil {
+		return err
+	}
+	log.Info("new leader ready")
+	for i := 1; i <= 3; i++ {
+		if err := c.transferPDAllocator(fmt.Sprintf("dc-%d", i)); err != nil {
+			return err
+		}
+	}
+	log.Info("new allocators ready")
+	if err := c.requestGlobalTSOErr(ctx); err != nil {
+		return err
+	}
+	return c.requestTSOs(ctx)
+}
+
+func (c *crossRegionClient) requestTSOs(ctx context.Context) error {
+	tsoCtx, tsoCancel := context.WithCancel(ctx)
+	defer tsoCancel()
+	tsoWg := &sync.WaitGroup{}
+	tsoWg.Add(4)
+	tsoErrCh := make(chan error, 4)
+	go c.requestTSO(tsoCtx, "global", tsoWg, tsoErrCh)
+	for i := 1; i <= 3; i++ {
+		go c.requestTSO(tsoCtx, fmt.Sprintf("dc-%v", i), tsoWg, tsoErrCh)
+	}
+	tsoWg.Wait()
+	if len(tsoErrCh) < 1 {
+		log.Info("requestTSOs success")
+		return nil
+	}
+	var tsoErrors []error
+	for len(tsoErrCh) > 0 {
+		tsoErr := <-tsoErrCh
+		tsoErrors = append(tsoErrors, tsoErr)
+	}
+	return util2.WrapErrors(tsoErrors)
+}
+
+func (c *crossRegionClient) requestTSO(ctx context.Context, dcLocation string, wg *sync.WaitGroup, errCh chan<- error) {
+	defer wg.Done()
+	if c.pdClient != nil {
+		for i := 0; i < c.TSORequests; i++ {
+			physical, logical, err := c.pdClient.GetLocalTS(ctx, dcLocation)
+			if err != nil {
+				log.Error("requestTSO failed", zap.String("dc-location", dcLocation), zap.Error(err))
+				errCh <- err
+				return
+			}
+			log.Info("request TSO", zap.String("dcLocation", dcLocation),
+				zap.Int64("physical", physical),
+				zap.Int64("logical", logical))
+		}
+	}
+}
+
+func (c *crossRegionClient) transferPDAllocator(dcLocation string) error {
```

Comment:
What about introducing the random way to transfer rather than the sequent way? Maybe it will cover more cases.

Reply:
That's a goood point. Currently I'd like to realize simple test for this case. In future we can inject more randomly action.